photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala

```diff
@@ -84,34 +87,40 @@ class CoordinateDescent(
     // Initialize the training scores
     var updatedScoresContainer = coordinates.map { case (coordinateId, coordinate) =>
       val updatedScores = coordinate.score(updatedGAMEModel.getModel(coordinateId).get)
-      updatedScores.setName(s"Initialized training scores with coordinateId $coordinateId")
-          .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
+      updatedScores
+        .setName(s"Initialized training scores with coordinateId $coordinateId")
+        .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
       (coordinateId, updatedScores)
     }.toMap
 
     // Initialize the validating scores
-    var validatingScoresContainerOption = validatingDataAndEvaluatorsOption.map { case (validatingData, _) =>
-      val validatingScoresContainer = coordinates.map { case (coordinateId, _) =>
+    val validatingScoresContainerOption = validatingDataAndEvaluatorsOption.map { case (validatingData, _) =>
+      coordinates.map { case (coordinateId, _) =>
         val updatedModel = updatedGAMEModel.getModel(coordinateId).get
         val validatingScores = updatedModel.score(validatingData)
-        validatingScores.setName(s"Initialized validating scores with coordinateId $coordinateId")
-            .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
+        validatingScores
+          .setName(s"Initialized validating scores with coordinateId $coordinateId")
+          .persistRDD(StorageLevel.INFREQUENT_REUSE_RDD_STORAGE_LEVEL)
         (coordinateId, validatingScores)
       }.toMap
-      validatingScoresContainer
     }
 
     // Initialize the regularization term value
-    var regularizationTermValueContainer = coordinates
-      .map { case (coordinateId, coordinate) =>
+    var regularizationTermValueContainer = coordinates.map { case (coordinateId, coordinate) =>
         val updatedModel = updatedGAMEModel.getModel(coordinateId).get
         (coordinateId, coordinate.computeRegularizationTermValue(updatedModel))
-      }
-      .toMap
+      }.toMap
+
+    // This will track the "best" model according to the first evaluation function chosen by the user.
+    // If the user did not specify any evaluation function, this var will be None.
+    var bestModel: Option[GAMEModel] = None
+    var bestEval: Option[Double] = None
```

Comment:
Either way is fine with me. I am not sure whether it makes a big difference in terms of readability, correctness or performance. We are probably not taking up the proposed change below though, as it allows information about the validation to percolate into the training. So we still need a bestModel var, and this proposed change feels less attractive to me. I'll punt on it for now.